### PR TITLE
Validate username password type when logging in, and validate email when reseting password

### LINKED ISSUE #2678: Security: NoSQL injection when loging in with username and password
### Issue Description

Describe your issue in as much detail as possible.

Regex query can be set into `req.body.username` if attacker use GET method with request body instead of URL query, which will lead a NoSQL injection.
https://github.com/ParsePlatform/parse-server/blob/master/src/Routers/UsersRouter.js#L86

### Steps to reproduce

Please include a detailed list of steps that reproduce the issue. Include curl commands when applicable.

1. Assume a user has registered with weak password like
```
username : user001@domain.com
password : 000000
```

2. If you login with posting request body instead of url parameter like <br>
`POST`
```
{
    "_method" : "GET", 
    "username" : {"$regex":"domain.com"},
    "password" : "000000"
}
```
which `{"$regex":"domain.com"}` will be parsed as a object in `req.body.username` at https://github.com/ParsePlatform/parse-server/blob/master/src/Routers/UsersRouter.js#L86

#### Results

If incidentally the first user found by database is `user001@domain.com`, login will be successful.

Of course the probability is very low in this case above but maybe it is possible to be abused in much more dangerous ways.

#### Solution?

- Escape the parameter in DB query

### PR DESCRIPTION
**Pull request by bamwang**
resolve #2678

### CODE REVIEW

**Comment by flovilmart** [comment]
Thanks @bamwang for this! Can you add a few tests around that so we don't introduce regressions in the future?

**Comment by bamwang** [comment]
I've just checked the Parse-SDK-JS which is used in the test, it seems that the validation is already embed in the logic (https://github.com/ParsePlatform/Parse-SDK-JS/blob/6c00cc825e0f34a88b010ea90da0c89e1a6f7b4b/src/ParseUser.js#L588-L607) or by flowtype.
It seems it is impossible to add tests with Parse-SDK-JS, should I add tests with pure http request?
Do you have any advice?

**Comment by flovilmart** [comment]
yep go with pure HTTP requests, you can use the `request-promise`  module for that. examples here: https://github.com/ParsePlatform/parse-server/blob/9ab488b6a02268511cbc73fbd5d6a1ef161c029b/spec/ParseUser.spec.js#L2509

**Comment by bamwang** [comment]
Great! Thanks for that. I will add tests soon.

**Comment by flovilmart** [comment]
np! Thanks for your contributions!

**Comment by bamwang** [comment]
Should I squash these two commit?

**Comment by flovilmart** [comment]
That looks good! Github squashes for us when we merge! 

Thanks again for this contribution!

File: spec/ParseUser.spec.js [hunk 1]
```diff
@@ -74,6 +74,58 @@ describe('Parse.User testing', () => {
...
+        }).catch((err) => {
+          expect(err.statusCode).toBe(404);
+          expect(err.message).toMatch('{"code":101,"error":"Invalid username/password."}');
+          done();
+        });
+      },
+    });
+  });
+
+  it('user login with non-string username with REST API', (done) => {
```
Comment (cherukumilli) [inline spec/ParseUser.spec.js]: @bamwang 
@flovilmart 
Minor correct to the test case title might be needed...

Should this say 
**user login with non-string `password` with REST API**

Comment (flovilmart) [inline spec/ParseUser.spec.js]: I didn't catch it sorry 👎 for @flovilmart

**Pull request merged**